File: vue/src/service/contact.service.js
import {api} from "../api/axiosInstance";

export const createContact = async (name , email, phone) => {
  try {
    const response = await api.post('/contacts', {
      name,
      email,
      phone
    },
    {
      headers: {
        'Authorization': `Bearer ${localStorage.getItem('token')}`
      }
    }
  );
    return response.data.data;
  } catch (error) {
    if (error.response) {
      throw new Error(error.response.data.message || 'Failed to create contact');
    } else {
      throw new Error('Network error');
    }
  }
}

export const getContacts = async () => {
  try {
    const response = await api.get('/contacts', {
      headers: {
        'Authorization': `Bearer ${localStorage.getItem('token')}`
      }
    });
    return response.data.data;
  } catch (error) {
    if (error.response) {
      throw new Error(error.response.data.message || 'Failed to fetch contacts');
    } else {
      throw new Error('Network error');
    }
  }
}

export const deleteContact = async (contactId) => {
  try {
    const response = await api.delete(`/contacts/${contactId}`, {
      headers: {
        'Authorization': `Bearer ${localStorage.getItem('token')}`
      }
    });
    return response.data.data;
  } catch (error) {
    if (error.response) {
      throw new Error(error.response.data.message || 'Failed to delete contact');
    } else {
      throw new Error('Network error');
    }
  }
}

export const updateContact = async (contactId, name, email, phone) => {
  try {
    const response = await api.put(`/contacts/${contactId}`, {
      name,
      email,
      phone
    },
    {
      headers: {
        'Authorization': `Bearer ${localStorage.getItem('token')}`
      }
    }
  );
    return response.data.data;
  } catch (error) {
    if (error.response) {
      throw new Error(error.response.data.message || 'Failed to update contact');
    } else {
      throw new Error('Network error');
    }
  }
}

File: blazor/Services/ContactService.cs
using System.Net.Http.Json;
using System.Text.Json.Serialization;
using rcontacts.Models;

namespace rcontacts.Services;

/// <summary>
/// Berisi CRUD operations untuk kontak
/// </summary>
public class ContactService
{
    private readonly ApiClient _api;

    public ContactService(ApiClient api)
    {
        _api = api;
    }

    /// <summary>
    /// GET /contacts
    /// </summary>
    public async Task<GetAllContactsResponse> GetContacts()
    {
        await _api.SetAuthHeader();
        var response = await _api.Client.GetAsync("/api/v2/contacts");

        if (!response.IsSuccessStatusCode)
        {
            var error = await response.Content.ReadFromJsonAsync<ServiceErrorResponse>();
            throw new Exception(error?.Message ?? "Gagal memuat kontak");
        }

        var result = await response.Content.ReadFromJsonAsync<GetAllContactsResponse>();
        return result ?? new GetAllContactsResponse();
    }

    /// <summary>
    /// POST /contacts
    /// </summary>
    public async Task CreateContact(string name, string email, string phone)
    {
        await _api.SetAuthHeader();
        var response = await _api.Client.PostAsJsonAsync("/api/v2/contacts", new { name, email, phone });

        if (!response.IsSuccessStatusCode)
        {
            var error = await response.Content.ReadFromJsonAsync<ServiceErrorResponse>();
            throw new Exception(error?.Message ?? "Gagal menambah kontak");
        }
    }

    /// <summary>
    /// PUT /contacts/:id
    /// </summary>
    public async Task UpdateContact(string id, string name, string email, string phone)
    {
        await _api.SetAuthHeader();
        var response = await _api.Client.PutAsJsonAsync($"/api/v2/contacts/{id}", new { name, email, phone });

        if (!response.IsSuccessStatusCode)
        {
            var error = await response.Content.ReadFromJsonAsync<ServiceErrorResponse>();
            throw new Exception(error?.Message ?? "Gagal memperbarui kontak");
        }
    }

    /// <summary>
    /// DELETE /contacts/:id
    /// </summary>
    public async Task DeleteContact(string id)
    {
        await _api.SetAuthHeader();
        var response = await _api.Client.DeleteAsync($"/api/v2/contacts/{id}");

        if (!response.IsSuccessStatusCode)
        {
            var error = await response.Content.ReadFromJsonAsync<ServiceErrorResponse>();
            throw new Exception(error?.Message ?? "Gagal menghapus kontak");
        }
    }
}

internal class ServiceErrorResponse
{
    public string? Message { get; set; }
}
